simple-acquaintances/data/datagen.py

```diff
@@ -0,0 +1,400 @@
+                truncnorm_gen = getTruncatedNormal(self.placesAffectingLocalThings[place][thing],
+                                                    self.localLikesVariance, 0, 1)
+                likeability[thing] += truncnorm_gen.rvs()
+            likeability[thing] /= len(person.lived)
+        
+        return likeability
+
+    def __flipCoin(self, bias = 0.5):
+        """
+        Return 1 or 0 based on a biased coin toss. Default bias = 0.5.
+        """
+        return 1 if np.random.random() < bias else 0
+
+    def __calculateSimilarity(self, p1, p2):
+        """
+        Return similarity measure between two persons (cosine similarity [0.0...1.0]).
+        """
+        person1 = np.array(p1.lived 
+                            + list(p1.globalLikes.values()) 
+                            + list(p1.localLikes.values())).reshape(1, -1)
+        person2 = np.array(p2.lived 
+                            + list(p2.globalLikes.values()) 
+                            + list(p2.localLikes.values())).reshape(1, -1)
+
+        return cosine_similarity(person1, person2)
+
+    def __generatePeople(self):
+        """
+        Generate list of people with their feature vectors filled in.
+        """
+        people = []
+        for i in range(self.numberOfPeople):
+            person = Person()
+
+            person.index = i
+
+            truncnormGenerator = getTruncatedNormal(self.placesLivedMean, self.placesLivedSD, 
+                                                    1, self.placesLivedUpper)
+            numberOfPlacesLived = int(truncnormGenerator.rvs())
+            person.lived = [0] * len(self.places)
+            placesLived = np.random.choice(self.places, numberOfPlacesLived, replace = False)
+            # One hot encoding of places lived by a person.
+            person.lived = [1 if i in placesLived else 0 for i in range(len(self.places))]
+
+            person.globalLikes = dict(zip(self.globalThings, 
+                                            np.random.uniform(0, 1, len(self.globalThings))))
+
+            person.localLikes = self.__getLikeability(person)
+            
+            people.append(person)
+
+        return people
+
+    def __generateKnows(self, people):
+        """
+        Return the knows truth matrix for all people in the dataset.
+        """
+        knowsMatrix = np.zeros((self.numberOfPeople, self.numberOfPeople))
+        for pair in combinations(people, 2):
+            similarity = self.__calculateSimilarity(pair[0], pair[1])
+            knows = self.__flipCoin(similarity)
+
+            # Set both (i,j) and (j,i) since P1 knows P2 => P2 knows P1.
+            knowsMatrix[pair[0].index][pair[1].index] = knows
+            knowsMatrix[pair[1].index][pair[0].index] = knows
+
+        return knowsMatrix
+
+    def __writeDataToFile(self, data, paths):
+        """
+        Write data to file.
+        """
+        if self.outputDir:
+            if self.outputDir[0] == '~':
```

Comment:
Nope.
This is real bad.
You should NEVER have to directly deal with things like this, you let the libraries who are versed in paths deal with it.

I suggest just not expanding the path, since your shell will automatically expand it and tilde should never be passed in a non-shell context.
But if you really want to handle it, then treat all paths the same:
```
self.outputDir = os.path.abspath(os.path.expanduser(self.outputDir))
```

Reply:
Also, handing all this on the edge (when you first get the path), not during the internals.